app/stable_appconfig.go

```diff
@@ -59,7 +68,88 @@ func setCustomOrderEndBlocker() []string {
 	return []string{}
 }
 
-func (app *RegenApp) registerUpgradeHandlers() {}
+func (app *RegenApp) registerUpgradeHandlers() {
+
+	// mainnet upgrade handler
+	const upgradeName = "v4.0.0"
+	app.UpgradeKeeper.SetUpgradeHandler(upgradeName, func(ctx sdk.Context, plan upgradetypes.Plan, fromVM module.VersionMap) (module.VersionMap, error) {
+		// run state migrations for sdk modules
+		toVersion, err := app.mm.RunMigrations(ctx, app.configurator, fromVM)
+		if err != nil {
+			return nil, err
+		}
+
+		// run x/ecocredit state migrations
+		if err := app.smm.RunMigrations(ctx, app.AppCodec()); err != nil {
+			return nil, err
+		}
+		toVersion[ecocredit.ModuleName] = ecocreditmodule.Module{}.ConsensusVersion()
+
+		// recover funds for community member (regen-1 governance proposal #11)
+		if ctx.ChainID() == "regen-1" {
```

Comment:
do we really need this check here? What if we want to run a testnet off-of mainnet candidate and test this

Reply:
Without the check, we would run recover funds on redwood. If you want to test on a testnet, you would need to update here and the check in `migrations/v3/patch.go`. We don't want to run recover funds if we don't need to.